scenario/smartsearch.go

```diff
@@ -0,0 +1,350 @@
+}
+
+func logSearchResult(logEntry *logger.LogEntry, id int, searchResult *enigma.SearchResult) {
+	if !logEntry.ShouldLogDebug() {
+		return
+	}
+	hasResult := false
+	for sgIdx, sga := range searchResult.SearchGroupArray {
+		for _, item := range sga.Items {
+			searchTermsMatched := []string{}
+			for _, stIndex := range item.SearchTermsMatched {
+				searchTermsMatched = append(searchTermsMatched, searchResult.SearchTerms[stIndex])
+			}
+			matches := []string{}
+			for _, match := range item.ItemMatches {
+				matches = append(matches, match.Text)
+			}
+			hasResult = true
+			logEntry.LogDebugf(`search%d result group%d matching term%s %s in %s %s: %s`, id, sgIdx+1,
+				ternary(len(searchTermsMatched) != 1, "s", ""), quoteList(searchTermsMatched),
+				item.ItemType, quote(item.Identifier), quoteList(matches))
+		}
+	}
+	if !hasResult {
+		logEntry.LogDebugf("search%d result: NONE", id)
+	}
+}
+
+var spaceRegex = regexp.MustCompile(`\s+`)
+
+func parseSearchTerms(s string) []string {
+	searchTerms := []string{}
+	sb := &strings.Builder{}
+	quoted := false
+	escaped := false
+	addSearchTerm := func() {
+		defer sb.Reset()
+		searchTerm := sb.String()
+		searchTerm = strings.TrimSpace(searchTerm)
+		searchTerm = spaceRegex.ReplaceAllString(searchTerm, " ")
+		if searchTerm != "" {
+			searchTerms = append(searchTerms, searchTerm)
+		}
+	}
+	for _, r := range s {
+		if escaped {
+			sb.WriteRune(r)
+		} else if r == '\\' {
+			escaped = true
+			continue
+		} else if r == '"' {
+			quoted = !quoted
+		} else if !quoted && unicode.IsSpace(r) {
+			addSearchTerm()
+		} else {
+			sb.WriteRune(r)
+		}
+		escaped = false
+	}
+	addSearchTerm()
+	return searchTerms
+}
+
+func doSmartSearchRPCs(sessionState *session.State, actionState *action.State, appDoc *enigma.Doc, id int, searchText string) {
+	searchTerms := parseSearchTerms(searchText)
+	if sessionState.LogEntry.ShouldLogDebug() {
+		sessionState.LogEntry.LogDebugf(`search%d text %s becomes search terms: %s`, id, quote(searchText), quoteList(searchTerms))
+	}
+
+	sessionState.QueueRequest(func(ctx context.Context) error {
+		searchSuggestionResult, err := appDoc.SearchSuggest(
+			ctx,
+			&enigma.SearchCombinationOptions{},
+			searchTerms,
+		)
+		logSearchSuggestionResult(sessionState.LogEntry, id, searchSuggestionResult)
+		return err
+	}, actionState, true, "SearchSuggest call failed")
+
+	sessionState.QueueRequest(func(ctx context.Context) error {
+		searchResult, err := appDoc.SearchResults(
+			ctx,
+			&enigma.SearchCombinationOptions{
+				Context:      "CurrentSelections",
+				CharEncoding: "Utf16",
+			},
+			searchTerms,
+			&enigma.SearchPage{
+				Offset: 0,
+				Count:  5,
+				GroupOptions: []*enigma.SearchGroupOptions{
+					{
+						GroupType: "DatasetType",
+						Offset:    0,
+						Count:     intPtr(-1),
+					},
+				},
+				GroupItemOptions: []*enigma.SearchGroupItemOptions{{
+					GroupItemType: "Field",
+					Offset:        0,
+					Count:         intPtr(5),
```

Comment:
Why 5? Perhaps add a const or comment

Reply:
It looks constant in the client. I have not observed any other value.

I add comments on this too.